data/builder/file.go

```diff
@@ -0,0 +1,265 @@
+package builder
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/ipfs/go-cid"
+	chunk "github.com/ipfs/go-ipfs-chunker"
+	"github.com/ipfs/go-unixfsnode/data"
+	dagpb "github.com/ipld/go-codec-dagpb"
+	"github.com/ipld/go-ipld-prime"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	basicnode "github.com/ipld/go-ipld-prime/node/basic"
+	"github.com/multiformats/go-multicodec"
+	multihash "github.com/multiformats/go-multihash/core"
+
+	// raw needed for opening as bytes
+	_ "github.com/ipld/go-ipld-prime/codec/raw"
+)
+
+// BuildUnixFSFile creates a dag of ipld Nodes representing file data.
+// This recreates the functionality previously found in
+// github.com/ipfs/go-unixfs/importer/balanced, but tailored to the
+// go-unixfsnode & ipld-prime data layout of nodes.
+// We make some assumptions in building files with this builder to reduce
+// complexity, namely:
+// * we assume we are using CIDv1, which has implied that the leaf
+//   data nodes are stored as raw bytes.
+//   ref: https://github.com/ipfs/go-mfs/blob/1b1fd06cff048caabeddb02d4dbf22d2274c7971/file.go#L50
+func BuildUnixFSFile(r io.Reader, chunker string, ls *ipld.LinkSystem) (ipld.Link, uint64, error) {
+	s, err := chunk.FromString(r, chunker)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	var prev []ipld.Link
+	var prevLen []uint64
+	depth := 1
+	for {
+		root, size, err := fileTreeRecursive(depth, prev[:], prevLen[:], s, ls)
```

Comment:
nit: I think these `[:]` expressions are no-ops, since `prev` and `prevLen` are already slices.